### ISSUE TITLE
Device Crashes when Host Device goes to sleep/powers down 

### ISSUE THREAD

**Issue by LibertyBeta** [adafruit/Adafruit_CircuitPython_MacroPad#23]
So, odd bug: the device will crash when the host goes into sleep/powers down.

The Display reads "USBerror: USB busy". Doesn't matter if USB memory is set to readonly or not, if usb_cdc is enabled, etc. It seems that any interactions with memory in the main loop while the Host device is in these power states causes the USB issue. 

Code can be found here: https://github.com/LibertyBeta/macro_pady

**Comment by FoamyGuy** [comment]
Does it crash immediately when the host device goes to sleep or powers off? or only when you press a button that is attempting to send shortcut keys to the host system? Or some other time?

**Comment by LibertyBeta** [comment]
It crashes almost instantly. No interaction is required. I haven't tested
if interaction makes it crash sooner.

**Comment by FoamyGuy** [comment]
Can you post the full stack trace that you're seeing? I'm curious to see which lines are initially raising the error.

**Comment by LibertyBeta** [comment]
Still trying to find a good way to dump the trace as it's when the host is
powered down.  Any ideas are welcome.

**Comment by LibertyBeta** [comment]
So, after MUCH tracing, looks like the issue is with a `macropad.keyboard.release_all()` call fouling things up.

**Comment by dhalbert** [comment]
`release_all()` sends a report with all the values set to value (nothing is pressed).

**Comment by LibertyBeta** [comment]
Yep, but that's what's crashing when the host is disconnected(IE, shutdown).

**Comment by hathach** [comment]
maybe this could help, https://github.com/hathach/tinyusb/pull/1020 would you mind switching the lib/tinyusb to branch to tes with https://github.com/hathach/tinyusb/tree/rp2040-disconnect-suspend

**Comment by rpavlik** [comment]
this is fascinating because I'd like to detect host sleep/powerdown in my code... I tried finding the right place for it in the circuitpython source this morning but had no luck. Right now I'm just doing a timeout based on last interaction.

**Comment by LibertyBeta** [comment]
https://github.com/hathach/tinyusb/pull/1020

This has been folding into TinyUSB, just waiting on news of how its going to be upstream into the main build. Still need details on this will interact with the Macropad.

**Comment by LibertyBeta** [comment]
Can confirm on most recent firmware that `supervisor.runtime.usb_connected` reports correctly now, but release, etc, still error. Should be a simple matter of having the lib check the supervisor to make sure there is an active connection before sending the keystrokes.

**Comment by mscreations** [comment]
I can confirm that changing the library to include a check for usb_connected at the beginning of all the release functions does seem to solve this issue. I put a check for it at the beginning of my main loop to gracefully shut down and stop doing anything that may call one of those functions which works well.

**Comment by ghost** [comment]
Is there any update on this bug? I am currently using the Macropad as a boot device to assist some boot time processes and in between bit locker and windows, If any key is pressed the board crashes.

**Comment by rpavlik** [comment]
I haven't seen this in a while. I would imagine that a new CircuitPython version sometime in the last year probably incorporated the fix.

**Comment by TheKitty** [comment]
A person has posted to the Adafruit guide feedback the following:

> My fix has been to include the following code at the top of the main while loop. This holds the program in place and prevents it reaching the bad line:
> 
> ```
> while supervisor.runtime.usb_connected == False:
> print("no USB connection")
> time.sleep(1)
> ```

**Comment by TheKitty** [comment]
I was asked to test this with current software. My environment:

- Windows 10, latest updates
- Adafruit MacroPad, updated to 9.2.8 stable with corresponding libraries
- Code in https://learn.adafruit.com/macropad-hotkeys/project-code pasted in as-is (no changes needed)

The code was tested and worked as normal on the MacroPad prior to PC sleep.
Windows was put to sleep
Power was cut to USB ports, macropad went dark
Exit Windows 10 from sleep
Macropad reset on USB port wake.

Changed Advanced Power Settings -> USB Settings -> USB selective suspend setting
from "Enabled" to "Disabled"
Retry sleep
Same result.

Possible conclusions:
- My machine goes into a deep sleep from the Power -> Sleep, not replicating the OPs conditions
- Changes in CircuitPython and TinyUSB in the interim possibly have alleviated this issue.

Suggestions:
- Retry with current 9.28 or higher CircuitPython and appropriate libraries
- If it still is not working, please repost in this thread. An interim method may be to use the code above noted by the feedback poster.
